### PR TITLE
Add URL-based language selector

### PR DESCRIPTION
**Pull request by jasonharrison**
* Add URL-based language selector, default to English
* Begin work on PT translations (mainly to show i18n logic)
  * @flamecoals, you can probably do the PT translations better than I can :-p
* Switch MessagePage to use i18n text
  * Update tests accordingly

### CODE REVIEW

File: src/App/App.tsx [hunk 1]
```diff
@@ -24,11 +24,17 @@ const withGA = (WrappedComponent: any) => {
...
 };
+
 const withI18n = (WrappedComponent: any) => {
-  class Wrapper extends React.Component<{}, {}> {
+  class Wrapper extends React.Component<any, {}> {
     render() {
-      registerLang('en', translations.en);
-      setCurrentLocale('en');
+      // load translations
+      for (const langCode of getLangCodes()) {
```
Comment (flamecoals) [inline src/App/App.tsx]: why the for loop?

File: src/App/App.tsx [hunk 2]
```diff
@@ -24,11 +24,17 @@ const withGA = (WrappedComponent: any) => {
...
+        registerLang(langCode, translations[langCode]);
+      }
+      // set language based on URL, default to English
+      const lang = this.props.match.params.locale || 'en';
```
Comment (flamecoals) [inline src/App/App.tsx]: you could use this for registerLang too

File: src/App/App.tsx [hunk 3]
```diff
@@ -39,6 +45,17 @@ const withWrappers = (WrappedComponent: any) => {
...
+const getLangCodes = () => {
+  return Object.keys(translations);
+};
+
+const generateBasePath = () => {
+  const langCodes = getLangCodes();
+  return `/:locale(${langCodes.join('|')})?`;
+};
+
+const base = generateBasePath();
```
Comment (flamecoals) [inline src/App/App.tsx]: remove base and generateBasePath()

File: src/App/App.tsx [hunk 4]
```diff
@@ -47,13 +64,13 @@ class Main extends React.Component<{}, {}> {
...
       <Switch>
-        <Route exact={true} path="/" component={withWrappers(Home)} />
-        <Route exact={true} path="/about" component={withWrappers(About)} />
-        <Route path="/g/:gameCode" exact={true} component={withWrappers(GameInfo)} />
-        <Route path="/g/:gameCode/:mode" exact={true} component={withWrappers(Game)} />
-        <Route path="/g/:gameCode/:mode/:aiLevel" exact={true} component={withWrappers(Game)} />
-        <Route path="/g/:gameCode/:mode/:matchCode/:playerID" exact={true} component={withWrappers(Game)} />
-        <Route component={withWrappers(getMessagePage('error', 'Not Found'))} />
+        <Route exact={true} path={base} component={withWrappers(Home)} />
+        <Route exact={true} path={`${base}/about`} component={withWrappers(About)} />
```
Comment (flamecoals) [inline src/App/App.tsx]: replace base with :locale([A-Za-z}{2})

File: src/App/MessagePageClass.test.tsx [hunk 5]
```diff
@@ -44,7 +44,7 @@ test('MessagePage link is hidden', () => {
   ).find(MessagePage);
   msg.setState({ linkHidden: true, startTime: 0 });
   (msg.instance() as MessagePage)._animate(125)();
-  expect(msg.html()).not.to.contain('Go Home');
+  expect(msg.html()).not.to.contain('messagePage.goHome');
```
Comment (flamecoals) [inline src/App/MessagePageClass.test.tsx]: why isnt it coming in english if it is the default language?

Comment (jasonharrison) [reply src/App/MessagePageClass.test.tsx]: Because there's no translations loaded.

**Pull request merged**